src/server/kernel/wserver.cpp

```diff
@@ -102,20 +104,23 @@ void WServerPrivate::init()
     }
 
     loop = wl_display_get_event_loop(display->handle());
-    int fd = wl_event_loop_get_fd(loop);
-
-    auto processWaylandEvents = [this] {
-        int ret = wl_event_loop_dispatch(loop, 0);
-        if (ret)
-            fprintf(stderr, "wl_event_loop_dispatch error: %d\n", ret);
-        wl_display_flush_clients(display->handle());
-    };
+    const int fd = wl_event_loop_get_fd(loop);
+    if (fd == -1) {
+        qCFatal(qLcWlrServer) << "Did not get the file descriptor for the event loop";
+    }
 
     sockNot.reset(new QSocketNotifier(fd, QSocketNotifier::Read));
-    QObject::connect(sockNot.get(), &QSocketNotifier::activated, q, processWaylandEvents);
+    QObject::connect(sockNot.get(), &QSocketNotifier::activated, q, [this] {
+        dispatchEvents();
+    });
 
-    QAbstractEventDispatcher *dispatcher = QThread::currentThread()->eventDispatcher();
-    QObject::connect(dispatcher, &QAbstractEventDispatcher::aboutToBlock, q, processWaylandEvents);
+    QAbstractEventDispatcher *dispatcher = QAbstractEventDispatcher::instance();
```

Comment:
The implementation is actually the same as QThread::currentThread()->eventDispatcher();